Add checklist navigation item to the sidebar

adamziel opened the pull request:
#### Changes proposed in this Pull Request

This PR adds a "Customer home" item to Calypso sidebar, linking to the checklist. It would finally give users a way to return to the checklist after getting lost in Calypso.

This is labeled "Customer Home" and not "Checklist" since the new customer home feature is going to replace this link in a couple of weeks and we want to make the transition a bit less confusing.

#### Testing instructions

Go to a site you are owner or admin of, the sidebar should contain the "Customer Home" link:
<img width="277" alt="Zrzut ekranu 2019-08-5 o 19 09 04" src="https://user-images.githubusercontent.com/205419/62482199-d8c07a80-b7b4-11e9-8fb2-079b1eb8bbf3.png">

Navigate to Customer Home, confirm the menu item is active:
<img width="273" alt="Zrzut ekranu 2019-08-5 o 19 09 00" src="https://user-images.githubusercontent.com/205419/62482198-d827e400-b7b4-11e9-8e9c-3659d963df82.png">

As a regular user (not a8c), go to a site you are not owner or admin of, confirm there is no "Customer Home" sidebar item:
<img width="269" alt="Zrzut ekranu 2019-08-5 o 18 43 32" src="https://user-images.githubusercontent.com/205419/62480742-f4298680-b7b0-11e9-993f-20bb5f36e018.png">

adamziel committed:
First stab at adding checklist to the sidebar

adamziel committed:
Revert development artifact

adamziel committed:
Add canCurrentUserUseChecklist check

adamziel committed:
Smooth out sharp edges

a8ci18n commented on client/my-sites/sidebar/sidebar.jsx:
```diff
@@ -155,6 +156,34 @@ export class MySitesSidebar extends Component {
+		return (
+			<SidebarItem
+				tipTarget="menus"
+				label={ translate( 'Checklist' ) }
```
Hi! I've found a possible matching string that has already been translated **2** times:
`translate( 'Launch Checklist' )` *ES Score: 7*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/add-checklist-to-sidebar)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

adamziel committed:
Rename Checklist to Customer Home

adamziel committed:
Use material icon for customer home

simison commented:
@jeffgolenski should this be hidden for jetpack sites or link to "my plan" page where the checklist is?

If you'd like to merge before anyone gets back about Jetpack, could you add `isJetpack` check to hide the menu? Thanks!

davemart-in commented:
A couple of questions/requests:

1) It looks like people will still be taken to the stats section by default? If that is the case, can we change the name of the link in the left nav from "Customer Home" to "Checklist" for now? We can always update it to "Home" for our release in 3 weeks. That, or we could just call it "Home" and take people there by default instead of the "Stats" screen.

2) I'm assuming that this will only be seen by new users (from the point of this PR going out)? Meaning existing customer (who signed up prior to this being released) won't see this left nav option?

3) I can't tell from the code, but is this new left nav option only shown to admins?

4) Are there any additional edge case signup scenarios to be mindful of outside of Jetpack sites? VIP sites? I'm trying to think of others.

5) So is the plan that if someone signs up and completes the checklist, will they just see the checklist congrats screen that promotes mobile for the next 3 weeks when they come to this section?

eliorivero commented:
What's this gap? Seen on Chrome/Mojave. It seems to be a focus outline or something. It's visible when I click on it.

<img width="291" alt="Captura de Pantalla 2019-08-05 a la(s) 15 24 35" src="https://user-images.githubusercontent.com/1041600/62486299-48267200-b795-11e9-9818-9ea537961177.png">

eliorivero reviewed (changes requested):
This is not working correctly on Jetpack sites. When I click on it, it takes me to the Plan page. 

I tried it on a wpcom site and the behavior is different: it takes me to the checklist when it's not completed, and takes me to the Congratulations! screen when it's completed. However, at the end, the Congratulations! screen asked me for a 2fa code. That was surprising and I'm not sure if it related to this. Maybe something in that page might have triggered it?

I expected that when I switch to a site, this would be the first page shown, not plans. I know, there's not much in this screen until we create the Customer Home.

I think that while this is essential for Customer Home, this shouldn't be released as it is now:
- functionality on Jetpack sites should be fixed.
- it should be hidden behind a flag. I had created a `customer-home` flag for our own development that could be helpful for this. Right now there's not much for users to gain from having this new menu item, until the Customer Home is ready.

michaeldcain commented:
> However, at the end, the Congratulations! screen asked me for a 2fa code

This is the current behavior of the congratulations screen when a user has 2fa active. The mobile app nudges require 2fa to pre-full the phone number prompts. It’s not ideal, but the majority of our new users will not have 2fa setup. 

> When I click on it, it takes me to the Plan page.

The Checklist for Jetpack sites is located on the plans page. I think that we should just hide this link for Jetpack sites. 

> it should be hidden behind a flag.

The point of this is to give a user easier access to the checklist while the customer home is still being built. The Customer Home will replace this. Maybe to remove any confusion, we can just call the item “Site Setup” or “Checklist”, and remove it when the Customer Home is ready. 

> On Aug 5, 2019, at 14:40, Elio Rivero <notifications@github.com> wrote:
> 
> @eliorivero requested changes on this pull request.
> 
> This is not working correctly on Jetpack sites. When I click on it, it takes me to the Plan page.
> 
> I tried it on a wpcom site and the behavior is different: it takes me to the checklist when it's not completed, and takes me to the Congratulations! screen when it's completed. However, at the end, the Congratulations! screen asked me for a 2fa code. That was surprising and I'm not sure if it related to this. Maybe something in that page might have triggered it?
> 
> I expected that when I switch to a site, this would be the first page shown, not plans. I know, there's not much in this screen until we create the Customer Home.
> 
> I think that while this is essential for Customer Home, this shouldn't be released as it is now:
> 
> functionality on Jetpack sites should be fixed.
> it should be hidden behind a flag. I had created a customer-home flag for our own development that could be helpful for this. Right now there's not much for users to gain from having this new menu item, until the Customer Home is ready.
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub, or mute the thread.

eliorivero commented:
> I think that we should just hide this link for Jetpack sites.

Good, hiding it for Jetpack sites would be perfect for now until CH is ready.

> Maybe to remove any confusion, we can just call the item “Site Setup” or “Checklist”,

Definitely, this was also suggested by Dave

> can we change the name of the link in the left nav from "Customer Home" to "Checklist" for now?

p-jackson committed:
Rename checklist nav item to "Checklist"

Was called Customer Home, but it wasn't going to be the default item so
wasn't the best name.

Removed the previously added home icons and switched to the checkmark
icon.

p-jackson committed:
Remove unneeded "stats" class from sidebar element

Presumably copy-pasted from the stats menu item

p-jackson committed:
Fixes visual glitch in menu item active ring

The ::after pseudo element used to fade out long sidebar labels was
causing a glitch in the Checklist item's active ring.

https://github.com/Automattic/wp-calypso/pull/35138#issuecomment-518346045

p-jackson committed:
Hide Checklist menu item for Jetpack sites

a8ci18n commented on client/my-sites/sidebar/sidebar.jsx:
```diff
@@ -155,6 +156,31 @@ export class MySitesSidebar extends Component {
+		return (
+			<SidebarItem
+				tipTarget="menus"
+				label={ translate( 'Checklist' ) }
```
Hi! I've found a possible matching string that has already been translated **2** times:
`translate( 'Launch Checklist' )` *ES Score: 7*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/add-checklist-to-sidebar)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

p-jackson commented:
- Renamed menu item to "Checklist" and changed icon
- Hidden menu for Jetpack sites

<img width="271" alt="image" src="https://user-images.githubusercontent.com/1500769/62496711-ebc36100-b82d-11e9-911f-c62a909dd9e2.png">

> What's this gap? Seen on Chrome/Mojave. It seems to be a focus outline or something. It's visible when I click on it.

@eliorivero seems to be some sort of CSS glitch that breaks the link's `:active` ring, caused by the `::after` element that handles text fade out.

<img width="271" alt="image" src="https://user-images.githubusercontent.com/1500769/62497018-d00c8a80-b82e-11e9-88bf-3aea1c8cda01.png">

Added a hack to fix the visual glitch.

p-jackson commented:
> I'm assuming that this will only be seen by new users

@davemart-in this PR currently shows the menu item to old and new users alike. I read that your plan for the new home page is to only show it to new customers, and for ease of migrating this link to the new one in 3 weeks it'd make sense for us to do the same.

On the other hand the checklist isn't something new to our users, we're fixing something that currently feels broken, not introducing a new feature. So I don't think this is quite the same case.

Maybe in 3 weeks this menu item could show either checklist _or_ the home link?

davemart-in commented:
> this PR currently shows the menu item to old and new users alike

@p-jackson Let's please not do this. There is no need to show this menu item to the hundreds of thousands of customers who have long since completed the checklist, or never saw it in the first place. Case in point, here's what I see when I visit the checklist from one of my (7 year) old accounts: 

![](https://d1jfzjx68gj8xs.cloudfront.net/items/2j0q430J001H3N102632/Screen%20Shot%202019-08-05%20at%207.00.24%20PM.png)

Showing this to existing users would only create confusion IMO. 

> On the other hand the checklist isn't something new to our users, we're fixing something that currently feels broken, not introducing a new feature. So I don't think this is quite the same case.

The checklist is important for new users only. I agree that it's hard to find now, but showing this to existing users feels even more broken to me.

davemart-in reviewed (changes requested):
One minor change request: Can we please update the temporary checklist left nav icon to match the one we use for Reader (See: https://cl.ly/e3c90c455e27)? The one we have in there now is a filled icon, which doesn't match well with the other lined icons: https://cl.ly/29250957637e

Note: Jetpack is filled, but that should be the outlier.

p-jackson committed:
Use material check_circle icon for checklist menu item

p-jackson commented:
> Can we please update the temporary checklist left nav icon to match the one we use for Reader

Done, looks much nicer :)

> There is no need to show this menu item to the hundreds of thousands of customers who have long since completed the checklist

This is a really good point. I'll just mention @adamziel and @michaeldcain for this. In the mean time I'll have a think about the right way to exclude old sites. Maybe the simplest way is based on site creation date, but it'd be nicer if it could be tied to this PR somehow instead since I don't know the exact date/time this'll ship 🤔

ramonjd commented:
> In the mean time I'll have a think about the right way to exclude old sites. Maybe the simplest way is based on site creation date

If it helps, we did do this once to test for checklist eligibility here: https://github.com/Automattic/wp-calypso/blob/212583cf098ebe47e361f7384350da14ed02a4e4/client/state/selectors/is-eligible-for-dotcom-checklist.js#L27

p-jackson committed:
Checklist menu hidden for existing - and other ineligible - sites

Not showing the checklist menu for existing sites because sites that
have been running successfully for years have no need for a Checklist
item to be added to the menu.

Also using the `isEligibleForDotcomChecklist()` selector to exclude
other sites that don't have a checklist e.g. sites where the
`design_type` option is `'store'`.

Add unit tests to try and set some of the conditions documented more
clearly.

p-jackson commented:
Thanks @ramonjd! That's highlighted some other cases where a site is ineligible for the checklist, so I've hidden the checklist menu item in those cases too by calling the `isEligibleForDotcomChecklist()` selector.

I've pushed a change so that the menu item only appears for sites created after **6 Aug 2019**.

andrewserong commented on client/state/sites/test/selectors.js:
```diff
@@ -3929,4 +3930,65 @@ describe( 'selectors', () => {
+			).toBe( false );
+		} );
+
+		test( 'should return flase for Jetpack site', () => {
```
Just a tiny typo here:
```suggestion
		test( 'should return false for Jetpack site', () => {
```

adamziel commented:
> So is the plan that if someone signs up and completes the checklist, will they just see the checklist congrats screen that promotes mobile for the next 3 weeks when they come to this section?

@davemart-in that's right. It's not perfect but it's better than the alternative. Plus I don't think it makes a lot of sense to hide the button once the checklist is completed.

Thank you SO MUCH for all your reviews. @p-jackson it's amazing how you jumped in and addressed all the feedback - thank you so much! Now - anyone dares to approve? :-)

davemart-in reviewed (approved):
LGTM from a design standpoint. 👍

spen committed:
Fix minor typo

Co-Authored-By: Andrew Serong <14988353+andrewserong@users.noreply.github.com>

spen reviewed (approved):
Code looks good to me and functionality works well! Glad to see a direct link to the checklist, thanks Adam!

spen commented:
Just pointing out that I see a couple of failing tests that look like they're related:

> Timed out waiting for element with css selector of '.checklist.main' to be present and displayed

https://circleci.com/gh/Automattic/wp-e2e-tests-for-branches/20890

adamziel commented:
One of them worked after rebuilding, I'm trying the other one

spen commented:
Nice! looks all clear! 🎉

eliorivero commented:
Works great. Tested with JP site and menu item was hidden. Tested with old wpcom account and it was hidden. Tested with new wpcom account and it was visible.
